### PR TITLE
add api for edit/delete cachegroup

### PR DESCRIPTION
**Pull request by zhilhuan**
Add RESTful API for edit/delete cache group.

Revise existing (create API) code to follow the API guidelines.

The APIs document is at https://docs.google.com/document/d/13GjIizusc7iUZq7TuVrsp8JZ2eEx52xUtoX9uRNzFgc/edit?usp=sharing

### CODE REVIEW

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 1]
```diff
@@ -256,83 +256,223 @@ Cache Group
 
   Role(s) Required: admin or oper
 
-  **Request Route Parameters**
+  **Request Parameters**
 
   +---------------------------------+----------+-------------------------------------------------------------------+
   | Name                            | Required | Description                                                       |
   +=================================+==========+===================================================================+
   | ``name``                        | yes      | The name of the Cache Group entry                                 |
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: all these fields are not required. check the database and update this table accordingly.

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 2]
```diff
@@ -256,83 +256,223 @@ Cache Group
...
+  +------------------------------------+--------+-------------------------------------------------------------------+
+  | ``typeName``                       | string | The type of Cache Group entry, "EDGE_LOC", "MID_LOC" or "ORG_LOC" |
+  +------------------------------------+--------+-------------------------------------------------------------------+
+  | ``lastUpdated``                    | string | The Time / Date this entry was last updated                       |
+  +------------------------------------+--------+-------------------------------------------------------------------+
 
   **Response Example** ::
 
     {
         "response": {
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: typically post, put, delete return a message as well. can you add an alerts node to this response? i'll show you how in a following comment...

File: traffic_ops/app/lib/API/Cachegroup.pm [hunk 3]
```diff
@@ -160,80 +160,275 @@ sub get_cachegroups {
...
-        $response->{secondary_parent_cachegroup_id} = $rs->secondary_parent_cachegroup_id;
-        $response->{type}        = $rs->type->id;
-        $response->{last_updated} = $rs->last_updated;
+    my ($response, $err1) = $self->get_cachegroup_by_id($insert->id);
+    if( defined($err1) ) {
+        return $self->alert(
+            { Error => $err1 }
+        );
+    }
+    return $self->success($response);
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Cachegroup.pm]: Can you change this to something like this? This will add an "alerts" node to the response which the api consumer can display if they so choose:

return $self->success($response, "Cachegroup successfully created: " . $name);

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 4]
```diff
@@ -256,83 +256,223 @@ Cache Group
...
+  +======+==========+====================================+
+  | id   | yes      | The id of the cache group to edit. |
+  +------+----------+------------------------------------+
+
+  **Request Parameters**
+
+  +---------------------------------+----------+-------------------------------------------------------------------+
+  | Name                            | Required | Description                                                       |
+  +=================================+==========+===================================================================+
+  | ``name``                        | no       | The name of the Cache Group entry                                 |
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: check the data schema. some of these are required.

File: traffic_ops/app/lib/API/Cachegroup.pm [hunk 5]
```diff
@@ -160,80 +160,275 @@ sub get_cachegroups {
...
+
+	&log( $self, "Update cachegroup with name:" . $update->name, "APICHANGE" );
+
+    my ($response, $err1) = $self->get_cachegroup_by_id($id);
+    if( defined($err1) ) {
+        return $self->alert(
+            { Error => $err1 }
+        );
     }
     return $self->success($response);
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Cachegroup.pm]: can you change to:

return $self->success($response, "Cachegroup was updated: " . $update->name);

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 6]
```diff
@@ -256,83 +256,223 @@ Cache Group
...
+  +-------------+--------+----------------------------------+
+  |  Parameter  |  Type  |           Description            |
+  +=============+========+==================================+
+  | ``alerts``  | array  | A collection of alert messages.  |
+  +-------------+--------+----------------------------------+
+  | ``>level``  | string | Success, info, warning or error. |
+  +-------------+--------+----------------------------------+
+  | ``>text``   | string | Alert message.                   |
+  +-------------+--------+----------------------------------+
+  | ``version`` | string |                                  |
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: get rid of version

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 7]
```diff
@@ -256,83 +256,223 @@ Cache Group
...
+  +-------------+--------+----------------------------------+
+
+  **Response Example** ::
+
+    {
+          "alerts": [
+                    {
+                            "level": "success",
+                            "text": "Cachegroup was deleted."
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: can you add the name of the cachegroup to the message?

File: traffic_ops/app/lib/API/Cachegroup.pm (hunk 3 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Cachegroup.pm]: can you change to:

return $self->success($response, "Cachegroup was created");

File: traffic_ops/app/t/api/1.2/cachegroup.t [hunk 8]
```diff
@@ -152,6 +152,84 @@ ok $t->post_ok('/api/1.2/cachegroups/9999/queue_update' =>  {Accept => 'applicat
     ->status_is(400)->or( sub { diag $t->tx->res->content->asset->{content}; } )
             , 'Does the queue_update api return?';
 
+my $cg_id = &get_cg_id('cache_group_edge');
```
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/cachegroup.t]: i'm getting a bunch of test failures when I `prove -qrp t`

t/api/1.2/cachegroup.t .................... 67/?
#   Failed test '200 OK'
#   at t/api/1.2/cachegroup.t line 223.
#          got: '404'
#     expected: '200'
# {"alerts":[{"level":"error","text":"Resource not found."}]}
#   Failed test '200 OK'
#   at t/api/1.2/cachegroup.t line 225.
#          got: '400'
#     expected: '200'
# {"alerts":[{"text":"Failed to delete cachegroup id = 14 has servers","level":"error"}]}
#   Failed test 'exact match for JSON Pointer "/alerts/0/level"'
#   at t/api/1.2/cachegroup.t line 225.
#          got: 'error'
#     expected: 'success'
#   Failed test 'exact match for JSON Pointer "/alerts/0/text"'
#   at t/api/1.2/cachegroup.t line 225.
#          got: 'Failed to delete cachegroup id = 14 has servers'
#     expected: 'Cachegroup was deleted.'
#   Failed test '404 Not Found'
#   at t/api/1.2/cachegroup.t line 228.
#          got: '400'
#     expected: '404'
# {"alerts":[{"level":"error","text":"Failed to delete cachegroup id = 14 has servers"}]}
#   Failed test '404 Not Found'
#   at t/api/1.2/cachegroup.t line 231.
#          got: '200'
#     expected: '404'

**Comment by mitchell852** [comment]
Tested POST, PUT and DELETE using curl...all seemed to work. address the comments and we can get this merged. thanks!

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 1 above)
Comment (zhilhuan) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: ok.

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 2 above)
Comment (zhilhuan) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: sure, updated.

File: traffic_ops/app/lib/API/Cachegroup.pm (hunk 3 above)
Comment (zhilhuan) [inline traffic_ops/app/lib/API/Cachegroup.pm]: sure, updated.

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 4 above)
Comment (zhilhuan) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: Sorry, I do not quite understand. If a field is not provide, the previous value in DB will be used. Why there will be some field required in update API?

File: traffic_ops/app/lib/API/Cachegroup.pm (hunk 5 above)
Comment (zhilhuan) [inline traffic_ops/app/lib/API/Cachegroup.pm]: sure, updated.

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 6 above)
Comment (zhilhuan) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: sure, updated.

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 7 above)
Comment (zhilhuan) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: sure, updated.

File: traffic_ops/app/t/api/1.2/cachegroup.t (hunk 8 above)
Comment (zhilhuan) [inline traffic_ops/app/t/api/1.2/cachegroup.t]: This is due to the "delete server" API not submitted in a separate PR. After merge master, this issue resolved.

**Comment by zhilhuan** [comment]
@mitchell852 Thanks a lot for your comments. I've updated all the comments, except for the PUT API doc, please check my response inline. I've also merged master, so UT will pass.

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 4 above)
Comment (limited) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: I thought all fields were required in a PUT? We don't have any examples of it yet, but the API guidelines say PATCH should be used for partial updates

Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: @limited - right, a PUT is a replace (although we don't follow this rule in many instances in the codebase). So typically you'd first do a GET to get the cachegroup, then change the cachegroup fields you want to change, and then a PUT with the entire cachegroup object. So for a PUT all required fields should be validated on the server. If any are missing a 400 Bad Request ($self-alert()) should be returned.

Whereas a PATCH allows the request payload to contain only a subset of fields <-- just the ones you want to change.

Since it looks like most of our PUTs in the codebase are really PATCHs, I think it's ok to call this a PUT. In TO v2.0/3.0 we may enforce this a little stronger however.

Comment (zhilhuan) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: @mitchell852 Thanks, it is clear for me now. Then would you like me to enforce PUT rule in this PR, or wait for TO v2.0/3.0 with other PUT APIs?

Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: @zhilhuan - I would say yes. Check that all required cache group fields are provided in the request json. If any are missing return $self-alert( { "field 1" -> " is required",  "field 2" -> " is required"})

check out is_valid() of User.pm to see how all the required fields of a User were checked - https://github.com/Comcast/traffic_control/blob/master/traffic_ops/app/lib/API/User.pm#L317

Comment (zhilhuan) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: @mitchell852 OK, got it. Updated in the following commit.

**Comment by mitchell852** [comment]
Still getting a bunch of failed unit tests. 14 of them to be exact. Here's what I did:

`cd traffic_control/traffic_ops/app`
`db/admin.pl --env=test reset`
`prove -qrp t`

Test Summary Report
-------------------
t/api/1.2/cachegroup.t                  (Wstat: 3584 Tests: 116 Failed: 14)
  Failed tests:  38-39, 87-88, 91-93, 96-98, 101, 104-106
  Non-zero exit status: 14

**Comment by zhilhuan** [comment]
@mitchell852 To make unit test pass, we need to merge PR #1550 together. This is due to POST server body keys changed to camelCase in that PR. And cachegroup.t used POST server API.

The reason I put last commit eed3c7a "change cachegroup ut code due to POST server API json body keys changed to camelCase in another PR" here instead of PR #1550 is, to avoid the merge conflict of the file "cachegroup.t".

**Comment by mitchell852** [comment]
ok, let's wait for #1550 to get merged, then can you update this PR?

**Comment by zhilhuan** [comment]
Yes, after #1550 merged. I will merge master, then the unit test will pass. I've already verified this in my local code base.
